sdk/core/core-lro/src/poller.ts

```diff
@@ -0,0 +1,133 @@
+  protected readonly client: ServiceClient;
+  protected initialResponse?: HttpOperationResponse;
+  protected previousResponse?: HttpOperationResponse;
+  protected requestOptions: RequestOptionsBase | undefined;
+  public stopped: boolean;
+
+  constructor(options: PollerOptionalParameters) {
+    this.manual = Boolean(options.manual);
+    this.intervalInMs = Number(options.intervalInMs);
+    this.client = options.client;
+    this.initialResponse = options.initialResponse;
+    this.previousResponse = options.previousResponse;
+    this.requestOptions = options.requestOptions;
+    this.stopped = Boolean(options.stopped);
+
+    if (options.state) this._state = options.state;
+    if (this.manual) return;
+    this.poll();
+  }
+
+  protected abstract getStateFromResponse(
+    response: HttpOperationResponse
+  ): LongRunningOperationStates;
+
+  protected abstract async sendRequest(options?: RequestOptionsBase): Promise<HttpOperationResponse>;
+
+  protected async processResponse(response: HttpOperationResponse): Promise<void | Result> {
+    if (!this.initialResponse) this.initialResponse = response;
+    this.previousResponse = response;
+    this.emit("newResponse", response, this);
+    this.state = this.getStateFromResponse(response);
+    if (this.isDone()) {
+      const result = await this.finalRequest();
+      this.emit("done", result);
+      return result;
+    }
+  }
+
+  protected async poll(): Promise<void> {
+    if (this.manual) return;
+    try {
+      if (!this.initialResponse) {
+        await this.processResponse(await this.initialRequest(this.requestOptions));
+      }
+      while (!this.isDone()) {
+        const interval: number = this.intervalInMs || this.getInterval();
+        await delay(interval);
+        await this.processResponse(await this.sendRequest());
+      }
+    } catch(e) {
+      this.emit("pollError", e, this);
+      throw e;
+    }
+  }
+
+  public abstract async initialRequest(options?: RequestOptionsBase): Promise<HttpOperationResponse>;
+
+  public abstract async finalRequest(options?: RequestOptionsBase): Promise<Result>;
+
+  public abstract async cancelRequest(options?: RequestOptionsBase): Promise<void>;
+
+  public abstract isDone(): boolean;
+
+  public set state(newState: LongRunningOperationStates) {
+    const newState = this.state !== newState;
+    this._state = newState;
+    if (newState) {
+      this.emit("newState", newState, this);
+    }
+  }
+
+  public get state(): LongRunningOperationStates {
+    return this._state;
+  }
+
+  public getInterval(): number {
+    return this.intervalInMs;
+  }
+ 
+  public async pollOnce(options?: RequestOptionsBase): Promise<void | Result> {
+    if (!this.manual) throw new Error("Manual retries are disabled on this poller");
+    try {
+      if (!this.initialResponse) {
+        await this.processResponse(await this.initialRequest(this.requestOptions));
+      }
+      return await this.processResponse(await this.sendRequest());
```

Comment:
nit: I would prefer if the request creation was what happened / was awaited on conditionally and then the request was passed to a single call to processResponse instead of that being duplicated across two lines

Reply:
Both responses need to be processed by process response though 🤔

Reply:
oh yes I meant something like

```ts
let request;
if (!this.initialResponse) {
  request = await this.initialRequest(this.requestOptions);
} else {
  request = await this.sendRequest();
}
return await this.processResponse(request);
```

Reply:
Ahh, gotcha! Thank you!